### PR TITLE
Create first version of the cli code

### PR DESCRIPTION
**Pull request by arbulu89**
**I would like to know the opinion about the code, we will think if we need to create other repository or package to store it, but by now this is the easiest way to review the code. After that we will add the documentation and the tests**

I have created the first version of the shaptools cli.
The idea of this piece of code is to expose the most used hana and ha commands in a unique entry point (what we have right now in the debug.sh).

Basically it uses the shaptools code calls.

### CODE REVIEW

**Comment by diegoakechi** [comment]
@arbulu89  Looks great for me!

File: shaptools/cli.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+        hana_instance.create_user_key(
+            hana_args.key_name, hana_args.environment, hana_args.user_name,
+            hana_args.user_password, hana_args.database)
+    elif str_args == 'backup':
+        hana_instance.create_backup(
+            hana_args.database, hana_args.name, hana_args.key_name,
+            hana_args.user_name, hana_args.user_password)
+
+
+def run_sr_subcommands(hana_instance, sr_args, logger):
```
Comment (MalloZup) [inline shaptools/cli.py]: what is `sr`?

Comment (fmherschel) [reply shaptools/cli.py]: System Replication?

Comment (arbulu89) [reply shaptools/cli.py]: Yes, sr stands for System replication

Comment (diegoakechi) [reply shaptools/cli.py]: Exactly! Thanks @fmherschel

Comment (MalloZup) [reply shaptools/cli.py]: We would just maybe add comment for the acronym

Comment (arbulu89) [reply shaptools/cli.py]: @MalloZup There is a `help` output for that (`-h` option).

**Comment by arbulu89** [comment]
Ok, After having a positive feedback I will proceed with the UT and documentation

**Comment by arbulu89** [comment]
I have finished the code:
- Documentation added
- UT created

File: docs/SHAPCLI.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+```
+Configuration file or sid,instance and passwords parameters must be provided
+
+usage: shapcli [-h] [--verbosity VERBOSITY] [--remotely REMOTELY] [-c CONFIG]
+               [-s SID] [-i INSTANCE] [-p PASSWORD]
+               {hana,sr} ...
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --verbosity VERBOSITY
```
Comment (diegoakechi) [inline docs/SHAPCLI.md]: @arbulu89 Could we adopt `-v` and `-r` here too just for easier usage?

File: python-shaptools.spec [hunk 3]
```diff
@@ -67,5 +67,6 @@ API to expose SAP HANA functionalities
 %license LICENSE
 %endif
 %{python_sitelib}/*
+%{_bindir}/shapcli
```
Comment (diegoakechi) [inline python-shaptools.spec]: Based on the FHS, I think the right location for this is `/usr/bin` as it is not essential.

https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

File: python-shaptools.changes [hunk 4]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Wed Jul 17 09:34:22 UTC 2019 - Xabier Arbulu Insausti <xarbulu@suse.com>
+
+- Create package version 0.3.0
+- shapcli is provided to expose shaptools api methodsas command line tool
```
Comment (MalloZup) [inline python-shaptools.changes]: typo: methodsas -> method as

File: docs/SHAPCLI.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+Check how it works and help output running:
+
+```
+shapcli -h
+```
+
+This is the output:
+
+```
```
Comment (MalloZup) [inline docs/SHAPCLI.md]: having done some CLI in past, I'm not against to have the ouput of `shapcli -h` in README.md or docs, but the risk problem is that if we remove/add new things, this are potential places where we need to remember to update the doc. IMHO we could remove  and just refer to the `-h`

File: docs/SHAPCLI.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
+# SHAPCLI
+
+shapcli is an executable tool to use the api provided by shaptools. It wraps most of the commands and
+exposes them as a command line tool.
+
+In order to use the utility the `shaptools` library must be installed (either using `pip` or `rpm` package).
+
+## Motivation
```
Comment (MalloZup) [inline docs/SHAPCLI.md]: rationale as hint but ok also motivation

File: python-shaptools.spec (hunk 3 above)
Comment (arbulu89) [reply python-shaptools.spec]: %{_bindir} is installing the file in /usr/bin, isn't. At least this is how it works in my system using Leap 15

File: docs/SHAPCLI.md (hunk 5 above)
Comment (arbulu89) [reply docs/SHAPCLI.md]: Ok, agree. I will remove the outputs, just leaving the usage examples

File: python-shaptools.spec (hunk 3 above)
Comment (diegoakechi) [reply python-shaptools.spec]: @arbulu89 Hum... you are right, but the setup.py points to bin/shapcli, is that correct?

Comment (arbulu89) [reply python-shaptools.spec]: setup.py points to the file that will be moved to /usr/bin (so in our repo is stored in bin/shapcli)

File: docs/SHAPCLI.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+Here an example of how to create the configuration file: [config.json](shapcli.config.example)
+
+Here some examples:
+
+```
+shapcli -s sid -i 00 -p HANAPASSWORD hana version
+shapcli -c config.json hana version
+```
+
+Check how it works and help output running:
```
Comment (MalloZup) [inline docs/SHAPCLI.md]: We could reformulate this with: 

`for more information please see the help information` 

somethink like that , Check how it works is imho more  for internal usage :grin:

File: shaptools/shapcli.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
+"""
+Code to expose some useful methods using the command line
```
Comment (MalloZup) [inline shaptools/shapcli.py]: this docstring can be improved

File: shaptools/shapcli.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+import argparse
+import json
+
+from shaptools import hana
+
+PROG = 'shapcli'
+LOGGING_FORMAT = '%(message)s'
+
+
+class DecodedFormatter(logging.Formatter):
```
Comment (MalloZup) [inline shaptools/shapcli.py]: do we need a Class here? I mean imho I would just use a helper function without the `super` overloading if possible.

Comment (MalloZup) [inline shaptools/shapcli.py]: I was asking me if we need a Class here, and we could not achieving it in a more functional way without it.

I think that this is perhaps the more pythonic way but I was thinking if we could just use an helper function for setting the  logging format. 

It is more a curiosity from the functional world. I"m also ok to have this with class and super

Comment (arbulu89) [reply shaptools/shapcli.py]: I created a class because the standard formatter itself is an object/class (actually in python everything is an object...). The unique thing I needed was to reformat the captured logs to remote the annoying `b''` part from the logs. As far as I know this is how the python logging works.
I'm afraid that is not easy to apply functional programming techniques here hehe
